helix-core/src/main/java/org/apache/helix/task/TaskUtil.java

```diff
@@ -1043,23 +1043,33 @@ public static void purgeExpiredJobs(String workflow, WorkflowConfig workflowConf
    * @param dataProvider
    * @param manager
    */
-  public static void workflowGarbageCollection(WorkflowControllerDataProvider dataProvider,
+  public static void workflowGarbageCollection(final WorkflowControllerDataProvider dataProvider,
       final HelixManager manager) {
     // Garbage collections for conditions where workflow context exists but config is missing.
-    Map<String, ZNRecord> contexts = dataProvider.getContexts();
-    HelixDataAccessor accessor = manager.getHelixDataAccessor();
-    HelixPropertyStore<ZNRecord> propertyStore = manager.getHelixPropertyStore();
 
+    // toBeDeletedWorkflows is a set that contains the name of the workflows that their contexts
+    // should be deleted.
     Set<String> toBeDeletedWorkflows = new HashSet<>();
-    for (Map.Entry<String, ZNRecord> entry : contexts.entrySet()) {
-      if (entry.getValue() != null
-          && entry.getValue().getId().equals(TaskUtil.WORKFLOW_CONTEXT_KW)) {
-        if (dataProvider.getWorkflowConfig(entry.getKey()) == null) {
-          toBeDeletedWorkflows.add(entry.getKey());
+    try {
+      Set<String> existingWorkflowContexts = new HashSet<>(dataProvider.getContexts().keySet());
+      for (String entry : existingWorkflowContexts) {
+        if (entry != null) {
+          WorkflowConfig cfg = dataProvider.getWorkflowConfig(entry);
+          WorkflowContext ctx = dataProvider.getWorkflowContext(entry);
+          if (ctx != null && ctx.getId().equals(TaskUtil.WORKFLOW_CONTEXT_KW) && cfg == null) {
+            toBeDeletedWorkflows.add(entry);
+          }
         }
       }
+    } catch (Exception e) {
+      LOG.warn(
+          "Exception occurred while creating a list of all existing contexts with missing config!",
+          e);
     }
```

Comment:
Do you clearly understand why and where ConcurrentModificationException is happening? This looks like we're just masking it with a try-catch. I don't think that's a fix. Could we identify why and where clearly first and could you please update the description with your findings? Then this try-catch shouldn't be necessary.

Reply:
Yes. I tried several scenarios and for each scenario I used Jiajun's scripts which runs the test for 50 time. The most effective solution is the one that I proposed in this PR. Please have a look at this line: 

- Set<String> existingWorkflowContexts = new HashSet<>(dataProvider.getContexts().keySet());

In some minor cases (about 2 out of 50 runs) above line is the only line that can generate concurrent modification exception which I eliminated it with try-catch. The reason behind this is because while we want to get all of the existing contexts, the contextMap can be modified in the cache by other threads. As a result we will get concurrentMod exception. Please note that this part of the code runs asynchronously.
@narendly I don't have strong preference about this method and I would be happy if you can propose new way to get all of the context without hitting concurrent modification exception.

Reply:
If you make a deep copy of the set, you shouldn't need to wrap it with a try-catch since your async thread would be the only thread iterating over the data structure, correct? What I'm not understanding is why we need the try-catch.

Reply:
I don't believe deep copy will help in this case (it would help in the single threaded case where you want to remove elements from the map or list). The code generate concurrent modification in this function: map.keySet(). We cannot avoid this. How you are proposing to get the keys in the map in this scenario? We need to get the keys no matter what. Even the implementation of deep copying involves to loop over the elements and copy them one by one to the new map. Right? In this case if original map has been changes while copy operation is happening, we might still get concurrent modification exception.

Reply:
@alirezazamani  thanks for the explanation :) Now that makes a lot more sense.

One structural improvement I'd suggest:
1. Please only wrap the copy constructor with a the try-catch to make your intentions clear.
2. It would be good to explain this in the code as a javadoc comment (as to why it needs to be wrapped with a try-catch).
3. Your catch should return the function early.

Reply:
Done. Thanks @narendly for the comments.